libraries/rush-lib/src/cli/actions/DeployAction.ts

```diff
@@ -144,31 +161,40 @@ export class DeployAction extends BaseRushAction {
 
     const createArchiveOnly: boolean = this._createArchiveOnly.value;
 
-    let transformPackageJson: ((packageJson: IPackageJson) => IPackageJson) | undefined;
-    let pnpmInstallFolder: string | undefined;
+    /**
+     * Subspaces that will be involved in deploy process.
+     * Each subspace may have its own configurations
+     */
+    const subspaces: IExtractorSubspace[] = [];
 
     const rushConfigurationProject: RushConfigurationProject | undefined =
       this.rushConfiguration.getProjectByName(mainProjectName);
-     if (!rushConfigurationProject) { 
-       throw new Error(`The specified deployment project "${mainProjectName}" was not found in rush.json`); 
-     } 
-
-    const subspace: Subspace = rushConfigurationProject.subspace;
+    if (!rushConfigurationProject) {
+      throw new Error(`The specified deployment project "${mainProjectName}" was not found in rush.json`);
+    }
 
+    const projects: Set<RushConfigurationProject> = this._getDependencyProjects(rushConfigurationProject);
     if (this.rushConfiguration.packageManager === 'pnpm') {
-      const pnpmfileConfiguration: PnpmfileConfiguration = await PnpmfileConfiguration.initializeAsync(
-        this.rushConfiguration,
-        subspace
-      );
-      transformPackageJson = pnpmfileConfiguration.transform.bind(pnpmfileConfiguration);
-      if (!scenarioConfiguration.json.omitPnpmWorkaroundLinks) {
-        pnpmInstallFolder = subspace.getSubspaceTempFolder();
+      for (const project of projects) {
+        const pnpmfileConfiguration: PnpmfileConfiguration = await PnpmfileConfiguration.initializeAsync(
+          this.rushConfiguration,
+          project.subspace
+        );
+        const subspace: IExtractorSubspace = {
+          subspaceName: project.subspace.subspaceName,
+          transformPackageJson: pnpmfileConfiguration.transform.bind(pnpmfileConfiguration)
+        };
+
+        if (!scenarioConfiguration.json.omitPnpmWorkaroundLinks) {
+          subspace.pnpmInstallFolder = project.subspace.getSubspaceTempFolder();
+        }
+        subspaces.push(subspace);
```

Comment:
Assuming A, B lives in the same subspace. It feels like the current logic generates the same subspace twice. Am I understanding this correctly?

Reply:
Nice catch, I might change it to a Set to avoid that case.